synapse/storage/data_stores/main/events.py

```diff
@@ -1954,6 +1954,78 @@ def insert_labels_for_event_txn(
             ],
         )
 
+    def insert_event_expiry(self, event_id, expiry_ts):
+        """Save the expiry timestamp associated with a given event ID.
+
+        Args:
+            event_id (str): The event ID the expiry timestamp is associated with.
+            expiry_ts (int): The timestamp at which to expire (delete) the event.
+        """
+        return self._simple_insert(
+            table="event_expiry",
+            values={"event_id": event_id, "expiry_ts": expiry_ts},
+            desc="insert_event_expiry",
+        )
+
+    def delete_expired_event(self, event):
+        """Delete an event that has expired by replacing its entry in event_json with a
+        pruned version of its JSON representation, and delete its associated expiry
+        timestamp.
+
+        Args:
+             event (events.EventBase): The event to delete.
+        """
+        # Prune the event's dict then convert it to JSON.
+        pruned_json = encode_json(prune_event_dict(event.get_dict()))
```

Comment:
as per #synapse-dev: I think you should consider using the existing redaction mechanism, rather than reinventing this.

Reply:
As per #synapse-dev, it's faffy to do a redaction without getting the event's original content included in `redacted_because`, so we're going with censoring the event. However, we're using the same code for censoring as the `_censor_redactions` function.